tests/test_metadata.py

```diff
@@ -618,3 +620,120 @@ def test_disallowed_dynamic(self, field_name):
+                "MIT AND (Apache-2.0 OR BSD-2-Clause)",
+            ),
+            (
+                "MIT OR GPL-2.0-or-later OR (FSFUL AND BSD-2-Clause)",
+                "MIT OR GPL-2.0-or-later OR (FSFUL AND BSD-2-Clause)",
+            ),
+            (
+                "GPL-3.0-only WITH Classpath-exception-2.0 OR BSD-3-Clause",
+                "GPL-3.0-only WITH Classpath-exception-2.0 OR BSD-3-Clause",
+            ),
+            (
+                "LicenseRef-Special-License OR CC0-1.0 OR Unlicense",
+                "LicenseRef-Special-License OR CC0-1.0 OR Unlicense",
+            ),
+            ("mIt", "MIT"),
+            ("mit or apache-2.0", "MIT OR Apache-2.0"),
+            ("mit and apache-2.0", "MIT AND Apache-2.0"),
+            (
+                "gpl-2.0-or-later with bison-exception-2.2",
+                "GPL-2.0-or-later WITH Bison-exception-2.2",
+            ),
+            (
+                "mit or apache-2.0 and (bsd-3-clause or mpl-2.0)",
+                "MIT OR Apache-2.0 AND (BSD-3-Clause OR MPL-2.0)",
+            ),
+            ("mit and (apache-2.0+ or mpl-2.0+)", "MIT AND (Apache-2.0+ OR MPL-2.0+)"),
+            # Valid non-SPDX values
+            ("LicenseRef-Public-Domain", "LicenseRef-Public-Domain"),
+            ("licenseref-public-domain", "LicenseRef-public-domain"),
+            ("licenseref-proprietary", "LicenseRef-proprietary"),
+            ("LicenseRef-Proprietary", "LicenseRef-Proprietary"),
+            ("LicenseRef-Beerware-4.2", "LicenseRef-Beerware-4.2"),
+            ("licenseref-beerware-4.2", "LicenseRef-beerware-4.2"),
+            ("", None),
+            (None, None),
+        ],
+    )
+    def test_valid_license_expression(self, license_expression, expected):
+        meta = metadata.Metadata.from_raw(
+            {"license_expression": license_expression}, validate=False
+        )
+        assert meta.license_expression == expected
+
+    @pytest.mark.parametrize(
+        "license_expression",
+        [
+            "Use-it-after-midnight",
+            "Apache-2.0 OR 2-BSD-Clause",
+            "LicenseRef-License with spaces",
+            "LicenseRef-License_with_underscores",
+            "or",
+            "and",
+            "with",
+            "mit or",
+            "mit and",
+            "mit with",
+            "or mit",
+            "and mit",
+            "with mit",
+            "(mit",
+            "mit)",
+            "mit or or apache-2.0",
+            "mit or apache-2.0 (bsd-3-clause and MPL-2.0)",
+        ],
+    )
+    def test_invalid_license_expression(self, license_expression):
+        meta = metadata.Metadata.from_raw(
+            {"license_expression": license_expression}, validate=False
+        )
+
+        with pytest.raises(metadata.InvalidMetadata):
+            meta.license_expression  # noqa: B018
+
+    @pytest.mark.parametrize(
+        "license_files",
+        [
+            ["LICENSE.txt", "licenses/*"],
```

Comment:
While technically a literal `*` is a valid filename, this looks like a glob to me and globs are not intended here. Is this on purpose, to check the literal `*`?

At the same time, the `licenses/` prefix is technically valid (the license file would be in `*.dist-info/licenses/licenses`), it looks weird as a test case. Is this also on purpose to check that such a silly case is valid?

Reply:
I definitely didn't put as much thought into the License-File part yet, awaiting feedback. this specific test case was misinterpreted from the [section of the pep relating to pyproject.toml](https://peps.python.org/pep-0639/#add-license-files-key).

With a little more consideration, I don't think globs should be showing up here at all given that clients should hav already resolved them.

Reply:
d6b47d5